app/plugins/weasyl.ts

```diff
@@ -11,41 +12,128 @@ export class WeasylPlugin extends BaseSitePlugin {
         }
         let button = buttonSpan.parentElement;
 
-        // Get the URL
-        // Weasyl download URLs are of the format 
-        // https://cdn.weasyl.com/~[user]/submissions/[id]/[hash]/[user]-[filename.ext]?download
-        let url = button.getAttribute("href");
-        let urlObj = new URL(url);
+        // If we get here, we probably have a canonical link and a download
+        // button.
+
+        // The canonical link has at least two formats; handle them both.
+        // "visual", "literary", and "multimedia" submissions:
+        // https://www.weasyl.com/%7E[user]/submissions/[id]/[title]
+        // "character" submissions:
+        // https://www.weasyl.com/character/[id]/[title]
+        let canonicalLink = canonical.getAttribute("href");
+        logger.log("weasyl: canonical href ", canonicalLink);
+
+        // Split on slashes
+        let canonicalObj = new URL(canonicalLink);
+        let pathParts = canonicalObj.pathname.split('/');
 
-        // Get the username
-        let pathParts = urlObj.pathname.split('/'); // "/~user/..."
-        let username = pathParts[1] || "";
-        username = username.replace("~", "");
+        let username = "";
+        let submissionIDIndex = 0;
+        let filenameIndex = 0;
 
-        // Get the filename
-        let serviceFilename = pathParts.pop();
-        pathParts.pop(); //pop off the unneeded hash
-        let id = pathParts.pop();
+        if(canonicalLink.indexOf("/submissions/") > -1) {
+            // regular submission
+            logger.log("weasyl: regular submission");
 
-        // Strip off the username
-        let { filename, ext } = getFilenameParts(serviceFilename);
-        filename = filename.substring(username.length + 1);
+            // Get the username and kill leading '%7E'
+            username = pathParts[1] || "";
+            username = username.slice(3);
+            logger.log("weasyl: username ", username);
 
+            // Record locations of these pieces for later
+            submissionIDIndex = 3;
+            filenameIndex = 4;
+        }
+        else if(canonicalLink.indexOf("/character/") > -1) {
+            // character submission
+            logger.log("weasyl: character submission");
+
+            // Username isn't in canonical link, so get it from the title
+            // above the artwork, which is of the format
+            // <h1 id="detail-title" class="pad-left pad-right">CoolGuy <i>by</i>
+            //   <a class="username" href="/~exampleartist">ExampleArtist</a></h1>
+            let usernameLink = document.querySelector('#detail-title a.username');
+            username = usernameLink.getAttribute("href");
+
+            // Kill leading '/~'
+            username = username.slice(2);
+            logger.log("weasyl: username ", username);
+
+            // Record locations of these pieces for later
+            submissionIDIndex = 2;
+            filenameIndex = 3;
```

Comment:
Instead of storing array indices, I'd recommend moving the variable definitions for `id` and `filename` up near the top of the function and initializing them with an empty string, then set the actual values as you come across them in the array. Avoids this weird indirection at least.

Reply:
I guess I was looking at the existing code, which used a fixed array index for username, and the equivalent of fixed indices (via popping from the far end) for username, serviceFilename, and and id.  I was trying to show that you have to get things from different places in the URL, depending on the submission type.

But, once I set those indices, I only use them once to actually set the related variables, so it might make more sense to do it your way.